File: problems/016/016.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* Copyright 2021 Melwyn Francis Carlo */

int main()
{
    char n_str[500] = "1024";

    char products_array[500][500];

    int str_length  = 4;

    for (int i = 2; i <= 100; i++)
    {
        for (int j = (str_length - 1); j >= 0; j--)
        {
            char temp_product[4+1];
            const char temp_char[2] = { n_str[j] };
            sprintf(temp_product, "%04d", 1024 * atoi(temp_char));

            for (int k1 = 0; k1 <= j; k1++)
                products_array[j][k1] = '0';

            for (int k2 = (j + 1); k2 < (j + 5); k2++)
                products_array[j][k2] = temp_product[k2-j-1];

            for (int k3 = (j + 5); k3 < (str_length + 4); k3++)
                products_array[j][k3] = '0';
        }

        memset(n_str, 0, sizeof(n_str));

        n_str[str_length + 4 - 1] = '0';

        for (int k = (str_length + 4 - 1); k > 0; k--)
        {
            int temp_sum = 0;
            for (int j = 0; j < str_length; j++)
                temp_sum += products_array[j][k] - '0';

            temp_sum  += n_str[k] - '0';
            n_str[k]   = ((temp_sum % 10) + '0');
            n_str[k-1] = ((int)(temp_sum / 10) + '0');
        }

        str_length += 4;
    }

    int sum_of_digits = 0;
    for (int i = 0; i < str_length; i++)
        sum_of_digits += n_str[i] - '0';

    printf("%d\n", sum_of_digits);

    return 0;
}

File: problems/016/016.f90
PROGRAM F016

    ! Copyright 2021 Melwyn Francis Carlo

    IMPLICIT NONE

    CHARACTER, DIMENSION(500) :: N_STR

    INTEGER :: SUM_OF_DIGITS = 0

    CHARACTER (LEN=4) :: TEMP_PRODUCT

    CHARACTER, DIMENSION(500,500) :: PRODUCTS_ARRAY

    INTEGER :: STR_LENGTH = 4

    INTEGER :: I, J, K, K1, K2, K3, TEMP_SUM

    N_STR(1) = '1'
    N_STR(2) = '0'
    N_STR(3) = '2'
    N_STR(4) = '4'

    DO I = 2, 100

        DO J = STR_LENGTH, 1, -1

            WRITE(TEMP_PRODUCT, '(I0.4)') 1024 * (IACHAR(N_STR(J)) - IACHAR('0'))

            DO K1 = 1, J
                PRODUCTS_ARRAY(J, K1) = '0'
            END DO

            DO K2 = (J + 1), (J + 4)
                PRODUCTS_ARRAY(J, K2) = TEMP_PRODUCT(K2-J:K2-J)
            END DO

            DO K3 = (J + 5), (STR_LENGTH + 4)
                PRODUCTS_ARRAY(J, K3) = '0'
            END DO

        END DO

        N_STR = '0'

        DO K = (STR_LENGTH + 4), 2, -1

            TEMP_SUM = 0
            DO J = 1, STR_LENGTH
                TEMP_SUM = TEMP_SUM + IACHAR(PRODUCTS_ARRAY(J, K)) - IACHAR('0')
            END DO
            TEMP_SUM = TEMP_SUM + IACHAR(N_STR(K)) - IACHAR('0')

            N_STR(K)   = ACHAR(MOD(TEMP_SUM, 10) + IACHAR('0'))
            N_STR(K-1) = ACHAR(FLOOR(REAL(TEMP_SUM) / 10.0) + IACHAR('0'))

        END DO

        STR_LENGTH = STR_LENGTH + 4

    END DO

    DO I = 1, STR_LENGTH
        SUM_OF_DIGITS = SUM_OF_DIGITS + IACHAR(N_STR(I)) - IACHAR('0')
    END DO

    PRINT ('(I0)'), SUM_OF_DIGITS

END PROGRAM F016
